cli/config/config.go

```diff
@@ -45,6 +52,8 @@ func ContextStoreDir() string {
 
 // SetDir sets the directory the configuration file is stored in
 func SetDir(dir string) {
+	// trigger the sync.Once to synchronise with Dir()
+	initConfigDir.Do(func() {})
```

Comment:
Didn't look too closely how these are called, but if both `Dir` and `SetDir` are called at the same time from different goroutines, we still might have a race condition:


1. Thread X enters `SetDir`
```go
56: initConfigDir.Do(func() {})
```

2. Thread Y enters `Dir` 
```go
39: initConfigDir.Do(func() {
	configDir = os.Getenv("DOCKER_CONFIG")
		if configDir == "" {
			configDir = filepath.Join(homedir.Get(), configFileDir)
		}
	})
	return configDir
}
```

3. Thread X enters the `Do` and "locks" `initConfigDir` first.
4. Thread Y is blocked and waits for X to finish the `Do` block.
5. Thread X leaves the `initConfigDir` and unblocks Y, but doesn't enter the `57: configDir = filepath.Clean(dir)` yet
6. Thread Y executes `45: return configDir` and returns empty directory.

Reply:
Can be ignored if we definitely don't (and won't) call these concurrently though.

Reply:
I don't _think_ we would hit that case in our code, so I'm "fine" with it not 100% being race-free (per our discussion on slack; there's no deadlock, so not "worse" than before).

FWIW; I went through a couple of options;

- a mutex for the value
- using `sync/atomic.Value`

Both of those performed a lot less, and both are still somewhat tapering over the "big elephant in the room"; we shouldn't use a package-level variable for this.

So I'm also looking at removing this variable altogether in favour of this being a config for the client that's configured, as (looking at the code) that's where it's used, and I think the expectation is to set the location _for the client_ that's constructed; https://github.com/docker/cli/blob/a26e6011f076481679ebd003eebfc988e55f0295/cli/command/cli.go#L209-L221